azure-iot-device/tests/iothub/pipeline/test_mqtt_topic_iothub.py

```diff
@@ -0,0 +1,1014 @@
+                id="Malformed topic",
+            ),
+        ],
+    )
+    def test_invalid_input_topic(self, topic):
+        with pytest.raises(ValueError):
+            mqtt_topic_iothub.get_input_name_from_topic(topic)
+
+
+@pytest.mark.describe(".get_method_name_from_topic()")
+class TestGetMethodNameFromTopic(object):
+    @pytest.mark.it("Returns the method name from a method topic")
+    def test_valid_method_topic(self):
+        topic = "$iothub/methods/POST/fake_method/?$rid=1"
+        expected_method_name = "fake_method"
+
+        assert mqtt_topic_iothub.get_method_name_from_topic(topic) == expected_method_name
+
+    @pytest.mark.it("URL decodes the returned method name")
+    def test_url_decodes_value(self):
+        topic = "$iothub/methods/POST/fake%24method/?$rid=1"
+        expected_method_name = "fake$method"
+        assert mqtt_topic_iothub.get_method_name_from_topic(topic) == expected_method_name
+
+    @pytest.mark.it("Raises a ValueError if the provided topic is not a method topic")
+    @pytest.mark.parametrize(
+        "topic",
+        [
+            pytest.param("not a topic", id="Not a topic"),
+            pytest.param(
+                "devices/fake_device/modules/fake_module/inputs/fake_input",
+                id="Topic of wrong type",
+            ),
+            pytest.param("$iothub/methdos/POST/fake_method/?$rid=1", id="Malformed topic"),
+        ],
+    )
+    def test_invalid_method_topic(self, topic):
+        with pytest.raises(ValueError):
+            mqtt_topic_iothub.get_method_name_from_topic(topic)
+
+
+@pytest.mark.describe(".get_method_request_id_from_topic()")
+class TestGetMethodRequestIdFromTopic(object):
+    @pytest.mark.it("Returns the request id from a method topic")
+    def test_valid_method_topic(self):
+        topic = "$iothub/methods/POST/fake_method/?$rid=1"
+        expected_request_id = "1"
+
+        assert mqtt_topic_iothub.get_method_request_id_from_topic(topic) == expected_request_id
+
+    # NOTE: valid request ids shouldn't need to be URL decoded, but we do it for safety
+    # and consistency. As a result, this test covers request_id values that are not valid
+    @pytest.mark.it("URL decodes the returned value")
+    def test_url_decodes_value(self):
+        topic = "$iothub/methods/POST/fake_method/?$rid=fake%24request%2Fid"
+        expected_request_id = "fake$request/id"
+        assert mqtt_topic_iothub.get_method_request_id_from_topic(topic) == expected_request_id
+
+    @pytest.mark.it("Raises a ValueError if the provided topic is not a method topic")
+    @pytest.mark.parametrize(
+        "topic",
+        [
+            pytest.param("not a topic", id="Not a topic"),
+            pytest.param(
+                "devices/fake_device/modules/fake_module/inputs/fake_input",
+                id="Topic of wrong type",
+            ),
+            pytest.param("$iothub/methdos/POST/fake_method/?$rid=1", id="Malformed topic"),
```

Comment:
>pytest.param("$iothub/methdos/POST/fake_method/?$rid=1", id="Malformed topic"), [](start = 12, length = 79)

This doesn't look particularly malformed to me.

Reply:
`methdos` and not `methods` so malformed 

---
In reply to: [399807652](https://github.com/Azure/azure-iot-sdk-python/pull/514#discussion_r399807652) [](ancestors = 399807652)

Reply:
Yeah, it's subtle, but it is malformed with a typo.

---
In reply to: [399807652](https://github.com/Azure/azure-iot-sdk-python/pull/514#discussion_r399807652) [](ancestors = 399807652)